pkg/compact/compact.go

```diff
@@ -573,42 +575,80 @@ func (cg *Group) compact(ctx context.Context, dir string, comp tsdb.Compactor) (
-			return compID, halt(errors.Wrapf(err, "compact planned compaction for mixed groups. group: %s, planned block's group: %s", cg.Key(), GroupKey(*meta)))
+			return false, halt(errors.Wrapf(err, "compact planned compaction for mixed groups. group: %s, planned block's group: %s", cg.Key(), GroupKey(*meta)))
 		}
 
 		for _, s := range meta.Compaction.Sources {
 			if _, ok := uniqueSources[s]; ok {
-				return compID, halt(errors.Errorf("overlapping sources detected for plan %v", plan))
+				return false, halt(errors.Errorf("overlapping sources detected for plan %v", plan))
 			}
 			uniqueSources[s] = struct{}{}
 		}
 
 		id, err := ulid.Parse(filepath.Base(pdir))
 		if err != nil {
-			return compID, errors.Wrapf(err, "plan dir %s", pdir)
+			return false, errors.Wrapf(err, "plan dir %s", pdir)
 		}
 
 		if err := block.Download(ctx, cg.bkt, id, pdir); err != nil {
-			return compID, retry(errors.Wrapf(err, "download block %s", id))
+			return false, retry(errors.Wrapf(err, "download block %s", id))
 		}
 
 		// Ensure all input blocks are valid.
-		if err := block.VerifyIndex(filepath.Join(pdir, block.IndexFilename), meta.MinTime, meta.MaxTime); err != nil {
-			return compID, halt(errors.Wrapf(err, "invalid plan block %s", pdir))
+		stats, err := block.GatherIndexIssueStats(filepath.Join(pdir, block.IndexFilename), meta.MinTime, meta.MaxTime)
+		if err != nil {
+			return false, errors.Wrapf(err, "gather index issues for block %s", pdir)
+		}
+
+		if err = stats.CriticalErr(); err == nil {
+			return false, halt(errors.Wrapf(err, "invalid plan block %s", pdir))
+		}
+
+		if err = stats.Issue347OutsideChunksErr(); err != nil {
+			level.Info(cg.logger).Log("msg", "stopping compaction and repairing invalid plan block", "err", err.Error(), "broken_block_id", pdir)
+
+			tmpdir, err := ioutil.TempDir("", fmt.Sprintf("index-issue-347-block-%s-", id))
+			if err != nil {
+				return false, err
+			}
+			defer os.RemoveAll(tmpdir)
+
+			// We need to repair this block before doing compaction.
+			resid, err := block.Repair(tmpdir, id, block.IgnoreIssue347OutsideChunk)
+			if err != nil {
+				return false, errors.Wrapf(err, "repair failed for block %s", id)
+			}
+
+			// Verify repaired block before uploading it.
+			if err := block.VerifyIndex(filepath.Join(tmpdir, resid.String(), block.IndexFilename), meta.MinTime, meta.MaxTime); err != nil {
+				return false, errors.Wrapf(err, "repaired block is invalid %s", resid)
+			}
+
+			level.Info(cg.logger).Log("msg", "uploading repaired block", "newID", resid)
+			if err = block.Upload(ctx, cg.bkt, filepath.Join(tmpdir, resid.String())); err != nil {
+				return false, retry(errors.Wrapf(err, "upload of %s failed", resid))
+			}
+
+			level.Info(cg.logger).Log("msg", "deleting broken block", "id", pdir)
+			if err := block.Delete(ctx, cg.bkt, id); err != nil {
+				return false, errors.Wrapf(err, "deleting old block %s failed. You need to delete this block manually", id)
+			}
+
+			return false, nil
```

Comment:
shouldn't we continue with execution here?

Reply:
I think we can remove this `return` as we would want to continue with the rest of the `pdir := range plan`